src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -100,82 +114,26 @@ static IEnumerable<Diagnostic> GetDynamicallyAccessedMembersDiagnostics (ValueSe
-					(true, false) => DiagnosticId.DynamicallyAccessedMembersMismatchMethodReturnTypeTargetsThisParameter,
-					(false, true) => DiagnosticId.DynamicallyAccessedMembersMismatchThisParameterTargetsMethodReturnType,
-					(false, false) => DiagnosticId.DynamicallyAccessedMembersMismatchThisParameterTargetsThisParameter
-				},
-				(SymbolKind.Method, SymbolKind.Parameter) => source.IsMethodReturn
-					? DiagnosticId.DynamicallyAccessedMembersMismatchMethodReturnTypeTargetsParameter
-					: DiagnosticId.DynamicallyAccessedMembersMismatchThisParameterTargetsParameter,
-				(SymbolKind.TypeParameter, SymbolKind.Field) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsField,
-				(SymbolKind.TypeParameter, SymbolKind.Method) => target.IsMethodReturn
-					? DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsMethodReturnType
-					: DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsThisParameter,
-				(SymbolKind.TypeParameter, SymbolKind.Parameter) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsParameter,
-				(SymbolKind.TypeParameter, SymbolKind.TypeParameter) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsGenericParameter,
-				// AnnotatedSymbol never stores a NamedType (yet). NamedType will only be used for known types which aren't implemented yet.
-				(SymbolKind.NamedType, _) => throw new NotImplementedException (),
-				(_, SymbolKind.NamedType) => throw new NotImplementedException (),
-				_ => throw new NotImplementedException ()
-			};
-
-		static string[] GetDiagnosticArguments (SymbolValue source, SymbolValue target, string missingAnnotations)
-		{
-			var args = new List<string> ();
-			args.AddRange (GetDiagnosticArguments (target));
-			args.AddRange (GetDiagnosticArguments (source));
-			args.Add (missingAnnotations);
-			return args.ToArray ();
-		}
-
-		static IEnumerable<string> GetDiagnosticArguments (SymbolValue annotatedSymbol)
-		{
-			ISymbol symbol = annotatedSymbol.Source;
-			var args = new List<string> ();
-			args.AddRange (symbol.Kind switch {
-				SymbolKind.Parameter => new string[] { symbol.GetDisplayName (), symbol.ContainingSymbol.GetDisplayName () },
-				SymbolKind.NamedType => throw new NotImplementedException (),
-				SymbolKind.Field => new string[] { symbol.GetDisplayName () },
-				SymbolKind.Method => new string[] { symbol.GetDisplayName () },
-				SymbolKind.TypeParameter => new string[] { symbol.GetDisplayName (), symbol.ContainingSymbol.GetDisplayName () },
-				_ => throw new NotImplementedException ($"Unsupported source or target symbol {symbol}.")
-			});
+			(var diagnosticId, var diagnosticArguments) = Annotations.GetDiagnosticForAnnotationMismatch (sourceWithDynamicallyAccessedMembers, targetWithDynamicallyAccessedMembers, missingAnnotations);
 
-			return args;
+			yield return Diagnostic.Create (DiagnosticDescriptors.GetDiagnosticDescriptor (diagnosticId), location, diagnosticArguments);
```

Comment:
I'm not so sure anymore. Right now it's not needed, but going forward once it implements the case for known type:
`typeof(Foo).RequirePublicMethods()`

This will have to go over all public methods on Foo and check for DAM and RUC, and potentially generate several warnings. It would be really neat to have all of this wrapped in this single method so that the callers don't have to deal with this in any way.

Reply:
That's a good point, I was just a little surprised to notice that it only ever yields one value.